pkg/controllers/machine/lifecycle/launch_timeout.go

```diff
@@ -0,0 +1,60 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package lifecycle
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/metrics"
+)
+
+type LaunchTimeout struct {
+	clock      clock.Clock
+	kubeClient client.Client
+}
+
+// launchTimeout is a heuristic time that we expect to succeed with our cloudprovider.Create() call
+// If we don't succeed within this time, then we should delete and try again through some other mechanism
+const launchTimeout = time.Minute * 2
+
+func (l *LaunchTimeout) Reconcile(ctx context.Context, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	if machine.StatusConditions().GetCondition(v1alpha5.MachineLaunched).IsTrue() {
+		return reconcile.Result{}, nil
+	}
+	if machine.StatusConditions().GetCondition(v1alpha5.MachineLaunched) == nil {
+		return reconcile.Result{Requeue: true}, nil
+	}
+	if l.clock.Since(machine.StatusConditions().GetCondition(v1alpha5.MachineLaunched).LastTransitionTime.Inner.Time) < launchTimeout {
+		return reconcile.Result{RequeueAfter: launchTimeout - l.clock.Since(machine.StatusConditions().GetCondition(v1alpha5.MachineLaunched).LastTransitionTime.Inner.Time)}, nil
+	}
+	// Delete the machine if we believe the machine won't create
+	if err := l.kubeClient.Delete(ctx, machine); err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+	logging.FromContext(ctx).With("timeout", launchTimeout).Debugf("deleting machine since node hasn't created within creation timeout")
+	metrics.MachinesTerminatedCounter.With(prometheus.Labels{
```

Comment:
I'd argue that these should be exposed externally and we should just call the metric for launching a machine prefixed with `machines_` i.e. right now we have `nodes_created` and we are moving to `machines_created`. The reason for this is that these two things have a semantically different meaning. Even if we change the impl later, I think we can still refer to the cloudprovider machine as the "machine" so that if we no longer have the local CR, we can still track metrics when we launch and terminate. WDYT?

Reply:
Was thinking about the launch_timeout label not surviving. Machine terminated probably makes sense to be public.

Reply:
I'm going to look around and see how other projects handle alpha metrics. Have you seen a pattern of using an `internal_` or `alpha_` prefix in metrics that projects aren't sure about?

Reply:
Do you think that we need to promise that _every_ metric name won't change? What if we just had the metrics but didn't surface them in our docs so we imply that we don't support them

Reply:
We should define this in our upgrade guide.